---
title: "R Notebook"
output: html_notebook
---


```{r}
library(tidyverse)
```

```{r}
setwd("/Users/isabelserrano/Documents/Science/Analyses/Conplastic_Strains/files_and_analyses/genomewide_selection_scan/hnhs_per_freq_bin/")

outdir_files =  "/Users/isabelserrano/Documents/Science/Analyses/Conplastic_Strains/files_and_analyses/genomewide_selection_scan/hnhs_per_freq_bin/files/"

obs_file = "files/mut_type_counts_bins_assigned.txt"
obs = read.table(obs_file, header=TRUE, stringsAsFactors = FALSE)

annotations_file = "../annotating_variants/output/annotated_all_possible_variants.txt"
annotations = read.table(annotations_file, header=TRUE, stringsAsFactors = FALSE)

total_num_sites_file = "files/total_annotated_num_sites_per_gene.txt"
total_num_sites = read.table(total_num_sites_file, header=TRUE, stringsAsFactors = FALSE)

```

We need to change F to FVB in strain name for annotation in order to merge our obs and annotation files

```{r}
renamed_annotations = annotations %>%
  mutate(STRAIN = ifelse(STRAIN == "F", "FVB", STRAIN))

```

Goal: Calculate the numerator for our hN and hS measures - the count of nonsyn and syn mutations per gene

```{r}

annotated_obs = obs %>%
  select(STRAIN, TISSUE, AGE_BIN, BIN, START, MUT_TYPE, COND_ALT_ALLELE_COUNT_AT_POS) %>%
  #renaming this column so that we can merge our annotations and our vcf
  rename(POS = START) %>%
  separate(MUT_TYPE, into = c("REF", "ALT"), remove = FALSE) %>%
  left_join(renamed_annotations, by = c("STRAIN", "POS", "REF", "ALT")) %>%
  #the na values appear for non-protein coding regions
  drop_na()

```


Calculating the observed count of mutation types per gene for each conditionxbin: 

```{r}

annotated_obs_counts = annotated_obs %>%
  select(STRAIN, TISSUE, AGE_BIN, BIN, GENE, ANNOTATION, COND_ALT_ALLELE_COUNT_AT_POS) %>%
  group_by(STRAIN, TISSUE, AGE_BIN, BIN, GENE, ANNOTATION) %>%
  summarise(COUNT = sum(COND_ALT_ALLELE_COUNT_AT_POS))

```
Pivoting our table from long to wide format

```{r}

annotated_obs_counts_wide = annotated_obs_counts %>%
  pivot_wider(names_from = ANNOTATION, values_from = COUNT) %>%
  #we drop cases where there weren't any syn or nsyn counts because we won't be able to calculate the ratio in these cases
  drop_na() %>%
  rename(OBS_NONSYN_COUNTS = missense_variant, OBS_SYN_COUNTS = synonymous_variant)
```

Merging our denominator and numerator dfs:

```{r}

obs_hnhs = annotated_obs_counts_wide %>%
  left_join(total_num_sites, by = c("STRAIN", "TISSUE", "AGE_BIN", "GENE")) %>%
  select(STRAIN, TISSUE, AGE_BIN, BIN, GENE, OBS_NONSYN_COUNTS, OBS_SYN_COUNTS, TOTAL_NONSYN_SITE_COUNT, TOTAL_SYN_SITE_COUNT) %>%
  mutate(hN = OBS_NONSYN_COUNTS/TOTAL_NONSYN_SITE_COUNT, hS = OBS_SYN_COUNTS/TOTAL_SYN_SITE_COUNT) %>%
  mutate(OBS_RATIO = hN/hS)

```

Exporting the observed ratios:

```{r}
setwd("/Users/isabelserrano/Documents/Science/Analyses/Conplastic_Strains/files_and_analyses/genomewide_selection_scan/hnhs_per_freq_bin/")

write.table(obs_hnhs, file = paste(outdir_files,"/obs_hNhS_ratios_per_gene_per_bin.txt", sep = ""), sep = "\t", quote = F, row.names = T)

```

