---
title: "R Notebook"
output: html_notebook
---


```{r}
library(tidyverse)
library(ggplot2)
library(PNWColors)
```

```{r}
outdir_figures = "../figures/"
outdir_files = "../files/"

nonreversion_muts_file ="/Users/isabelserrano/Documents/Science/Analyses/Conplastic_Strains/files_and_analyses/input_files/somatic_mutations.vcf"
nonreversion_muts = read.table(nonreversion_muts_file, header=TRUE, stringsAsFactors = FALSE)

reversion_muts_file = "/Users/isabelserrano/Documents/Science/Analyses/Conplastic_Strains/files_and_analyses/input_files/haplotype_mutations.vcf"
reversion_muts = read.table(reversion_muts_file, header=TRUE, stringsAsFactors = FALSE)

```

Combine all our variants info -- both the files that contain haplotype sites and files that contain all other sites: 
```{r}
all_variants = rbind(nonreversion_muts, reversion_muts)
```

```{r}
rm(nonreversion_muts, reversion_muts)
```

Cluster 1: 5 haplotype sites spanning 104 bp

```{r}

cluster_1 = data.frame(START = c(15498, 15548, 15577, 15587, 15602), REF = c("T", "C", "A", "C", "C"), ALT = c("A", "T", "T", "T", "T"))

```

The contamination frequency is given by the B6/ALR/FVB allele (REF allele) divided by the read depth at that pos.
```{r}

cluster_1_cont = inner_join(all_variants, cluster_1, by = c("START", "REF", "ALT")) %>%
  filter(STRAIN == "NZB") %>%
  select(SAMPLE, START, REF, ALT, REF_ALLELE_DEPTH, READ_DEPTH_AT_POS) %>%
  mutate(CONT_EST = REF_ALLELE_DEPTH/READ_DEPTH_AT_POS) %>%
  mutate(CONT_FLAG = ifelse(CONT_EST == 0, "NO", "YES"))

```

We calculate the minimum frequency in a cluster. Assuming that a contaminated read has the REF allele at all sites in a cluster, the minimum frequency would denote the proportion of reads present as a result of contamination. 

```{r}
min_cont_cluster_1 = cluster_1_cont %>%
  group_by(SAMPLE) %>%
  summarise(MIN_CONT = min(CONT_EST)) %>%
  mutate(CLUSTER = "clust_1")
```

```{r}

nzb_cluster_1_cont = ggplot(cluster_1_cont, aes(x = START, y = CONT_EST, color = CONT_FLAG)) + 
  geom_point(size = 0.7) + 
  geom_text(data = min_cont_cluster_1, aes(x =15520, y = 0.001, label = formatC(MIN_CONT, digits = 2, format = "e")), size = 2, inherit.aes = FALSE) + 
  facet_wrap(.~SAMPLE, nrow = 6) + 
  theme_bw() + 
  ylab("Contamination Est.") + 
  xlab("Haplotype site") + 
  scale_x_continuous(breaks = c(15498, 15548, 15577, 15587, 15602)) + 
  scale_color_manual(name = "Contamination", values = c("black", "magenta")) + 
  theme(strip.background = element_blank(),
        axis.text.x = element_text(angle = 90))

pdf(paste(outdir_figures,"/nzb_cluster_1_cont.pdf",sep=""),width=7,height=5.5)
print(nzb_cluster_1_cont)
dev.off()

```

Cluster 2: 5 haplotype sites spanning 75 bp

```{r}

cluster_2 = data.frame(START = c(2765, 2766, 2797, 2813, 2839), REF = c("A", "T", "C", "T", "C"), ALT = c("G", "C", "T", "C", "T"))

```

```{r}

cluster_2_cont = inner_join(all_variants, cluster_2, by = c("START", "REF", "ALT")) %>%
  filter(STRAIN == "NZB") %>%
  select(SAMPLE, START, REF, ALT, REF_ALLELE_DEPTH, READ_DEPTH_AT_POS) %>%
  mutate(CONT_EST = REF_ALLELE_DEPTH/READ_DEPTH_AT_POS) %>%
  mutate(CONT_FLAG = ifelse(CONT_EST == 0, "NO", "YES"))

```

```{r}

min_cont_cluster_2 = cluster_2_cont %>%
  group_by(SAMPLE) %>%
  summarise(MIN_CONT = min(CONT_EST)) %>%
  mutate(CLUSTER = "clust_2")

```

```{r}

nzb_cluster_2_cont = ggplot(cluster_2_cont, aes(x = START, y = CONT_EST, color = CONT_FLAG)) + 
  geom_point(size = 0.7) + 
  geom_text(data = min_cont_cluster_2, aes(x =2775, y = 0.001, label = formatC(MIN_CONT, digits = 2, format = "e")), size = 2, inherit.aes = FALSE) + 
  facet_wrap(.~SAMPLE, nrow = 6) + 
  theme_bw() + 
  ylab("Contamination Est.") + 
  xlab("Haplotype site") + 
  scale_x_continuous(breaks = c(2765, 2766, 2797, 2813, 2839)) + 
  scale_color_manual(name = "Contamination", values = c("black", "magenta")) + 
  theme(strip.background = element_blank(),
        axis.text.x = element_text(angle = 90))

pdf(paste(outdir_figures,"/nzb_cluster_2_cont.pdf",sep=""),width=7,height=5.5)
print(nzb_cluster_2_cont)
dev.off()

```

cluster 3: 3 haplotype sites spanning 90 bp


```{r}

cluster_3 = data.frame(START = c(11842, 11845, 11932), REF = c("G", "C", "A"), ALT = c("A", "T", "C"))

```


```{r}

cluster_3_cont = inner_join(all_variants, cluster_3, by = c("START", "REF", "ALT")) %>%
  filter(STRAIN == "NZB") %>%
  select(SAMPLE, START, REF, REF_ALLELE_DEPTH, READ_DEPTH_AT_POS) %>%
  mutate(CONT_EST = REF_ALLELE_DEPTH/READ_DEPTH_AT_POS) %>%
  mutate(CONT_FLAG = ifelse(CONT_EST == 0, "NO", "YES"))

```


```{r}

min_cont_cluster_3 = cluster_3_cont %>%
  group_by(SAMPLE) %>%
  summarise(MIN_CONT = min(CONT_EST)) %>%
  mutate(CLUSTER = "clust_3")

```

```{r}

nzb_cluster_3_cont = ggplot(cluster_3_cont, aes(x = START, y = CONT_EST, color = CONT_FLAG)) + 
  geom_point(size = 0.7) + 
  geom_text(data = min_cont_cluster_3, aes(x = 11860, y = 0.001, label = formatC(MIN_CONT, digits = 2, format = "e")), size = 2, inherit.aes = FALSE) + 
  facet_wrap(.~SAMPLE, nrow = 6) + 
  theme_bw() + 
  ylab("Contamination Est.") + 
  xlab("Haplotype site") + 
  scale_x_continuous(breaks = c(11842, 11845, 11932)) + 
  scale_color_manual(name = "Contamination", values = c("black", "magenta")) + 
  theme(strip.background = element_blank(),
        axis.text.x = element_text(angle = 90))

pdf(paste(outdir_figures,"/nzb_cluster_3_cont.pdf",sep=""),width=7,height=5.5)
print(nzb_cluster_3_cont)
dev.off()

```


Cluster 4: 3 haplotype site spanning 56 bp

```{r}

cluster_4 = data.frame(START = c(13780, 13781, 13836), REF = c("A", "T", "A"), ALT = c("G", "C", "G"))

```


```{r}

cluster_4_cont = inner_join(all_variants, cluster_4, by = c("START", "REF", "ALT")) %>%
  filter(STRAIN == "NZB") %>%
  select(SAMPLE, START, REF, REF_ALLELE_DEPTH, READ_DEPTH_AT_POS) %>%
  mutate(CONT_EST = REF_ALLELE_DEPTH/READ_DEPTH_AT_POS) %>%
  mutate(CONT_FLAG = ifelse(CONT_EST == 0, "NO", "YES"))

```


```{r}

min_cont_cluster_4 = cluster_4_cont %>%
  group_by(SAMPLE) %>%
  summarise(MIN_CONT = min(CONT_EST)) %>%
  mutate(CLUSTER = "clust_4")

```

```{r}

nzb_cluster_4_cont = ggplot(cluster_4_cont, aes(x = START, y = CONT_EST, color = CONT_FLAG)) + 
  geom_point(size = 0.7) + 
  geom_text(data = min_cont_cluster_4, aes(x = 11860, y = 0.001, label = formatC(MIN_CONT, digits = 2, format = "e")), size = 2, inherit.aes = FALSE) + 
  facet_wrap(.~SAMPLE, nrow = 6) + 
  theme_bw() + 
  ylab("Contamination Est.") + 
  xlab("Haplotype site") + 
  scale_x_continuous(breaks =c(13780, 13781, 13836)) + 
  scale_color_manual(name = "Contamination", values = c("black", "magenta")) + 
  theme(strip.background = element_blank(),
        axis.text.x = element_text(angle = 90))

pdf(paste(outdir_figures,"/nzb_cluster_4_cont.pdf",sep=""),width=7,height=5.5)
print(nzb_cluster_3_cont)
dev.off()

```

Combining cluster contamination information
```{r}
cont_across_clusters_df = rbind(min_cont_cluster_1, min_cont_cluster_2, min_cont_cluster_3, min_cont_cluster_4)

```

Exporting the contamination across clusters

```{r}
write.table(cont_across_clusters_df, file = paste(outdir_files,"nzb_cluster_cont_est.txt", sep = ""), sep = "\t", quote = F, row.names = F)
```

Summary stats of contamination across the four clusters
```{r}
plotting_cont_range_nzb = cont_across_clusters_df %>%
  group_by(SAMPLE) %>%
  summarise(AVG_CONT_CLUST = mean(MIN_CONT), MAX = max(MIN_CONT), MIN = min(MIN_CONT))
```

Exporting the contamination per sample

```{r}
write.table(plotting_cont_range_nzb, file = paste(outdir_files,"nzb_sample_cont_est_range.txt", sep = ""), sep = "\t", quote = F, row.names = F)

```

Plotting the range of contamination est per sample

```{r}

nzb_cont_range = ggplot(plotting_cont_range_nzb, aes(x = AVG_CONT_CLUST, y = SAMPLE)) + 
  geom_point(size = 2.3, shape = 18) + 
  geom_point(data = cont_across_clusters_df, aes(x = MIN_CONT, y = SAMPLE, color = CLUSTER),inherit.aes = FALSE, size = 1.3) + 
  geom_segment(aes(x = MIN, xend = MAX, y = SAMPLE, yend = SAMPLE)) + 
  scale_color_manual(name = "Cluster", labels = c("cluster 1", "cluster 2", "cluster 3", "cluster 4"), values = c("#6e7cb9", "#7bbcd5", "#d0e2af","#f5db99")) + 
  theme_bw() + 
  xlab("Estimated Contamination") + 
  ylab("Sample")


pdf(paste(outdir_figures,"/nzb_cont_range.pdf",sep=""),width=7,height=5.5)
print(nzb_cont_range)
dev.off()

```

Now we want to calculate the NZB contamination per condition:

We need the following files:
```{r}
#average read depth file per sample
average_read_depth_file ="/Users/isabelserrano/Documents/Science/Analyses/Conplastic_Strains/files_and_analyses/experimental_design/files/sample_duplex_depth.txt"
average_read_depth = read.table(average_read_depth_file, header=TRUE, stringsAsFactors = FALSE)

#chr1 NUMT contamination estimates to compare 
nzb_numt_corr_file = "/Users/isabelserrano/Documents/Science/Analyses/Conplastic_Strains/files_and_analyses/reversions/files/corrected_reversion_counts.txt"

nzb_numt_corr = read.table(nzb_numt_corr_file, header=TRUE, stringsAsFactors = FALSE) %>%
  filter(STRAIN == "NZB") %>%
  select(STRAIN, AGE_BIN, TISSUE, CONT_PERC) %>%
  filter(CONT_PERC > 0) %>%
  unique() %>%
  mutate(COND_LABEL = paste(AGE_BIN, TISSUE, sep = "_")) %>%
  select(COND_LABEL, CONT_PERC) %>%
  rename(COND_CONT_EST = CONT_PERC) %>%
  mutate(TYPE = "NUMT_CORRECTION_FACTOR")


```

We calculate the contamination for each biological condition by using the maximum level of contamination to estimate the proportion of contaminated reads in each sample. We then sum the number of contaminated reads and the read depth at each position across samples in a condition. We use these aggregates to calculate the contamination est as the total number of contaminated reads divided by the total read depth at the position.

```{r}
cond_nzb_est = plotting_cont_range_nzb %>%
  left_join(average_read_depth, by = "SAMPLE") %>%
  select(SAMPLE, TISSUE, AGE_BIN, MAX, AVG_DUP_DEPTH) %>%
  mutate(CONT_READS = MAX*AVG_DUP_DEPTH) %>%
  select(TISSUE, AGE_BIN, AVG_DUP_DEPTH, CONT_READS) %>%
  group_by(TISSUE, AGE_BIN) %>%
  summarise(COND_DUP_DEPTH = sum(AVG_DUP_DEPTH), COND_CONT_READS = sum(CONT_READS)) %>%
  mutate(COND_CONT_EST = COND_CONT_READS/COND_DUP_DEPTH) %>%
  mutate(COND_LABEL = paste(AGE_BIN, TISSUE, sep = "_")) %>%
  ungroup() %>%
  select(COND_LABEL, COND_CONT_EST) %>%
  mutate(TYPE = "CONT_EST")
```

```{r}
nzb_numt_corr_comp_df = rbind(nzb_numt_corr, cond_nzb_est)
```

Comparing the contamination estimate to the applied chr1 NUMT correction:

```{r}
nzb_numt_corr_comp = ggplot(nzb_numt_corr_comp_df, aes(x = COND_CONT_EST, y = COND_LABEL, color = TYPE)) + 
 geom_point() + 
  theme_bw() + 
  xlab("Condition Contamination \nEst.") + 
  ylab("Condition") + 
  scale_color_manual(name = "Measures", values = c("black", "magenta"), labels = c("Contamination Est", "Applied NUMT correction \nfactor (chrM: 6394-11042)")) + 
  theme(axis.text.x = element_text(angle = 90), 
        axis.title.y = element_text(size = 13),
        legend.position = "none")

pdf(paste(outdir_figures,"/nzb_numt_corr_comp.pdf",sep=""),width=2.85,height=3)
print(nzb_numt_corr_comp)
dev.off()

pdf(paste(outdir_figures,"/leg_nzb_numt_corr_comp.pdf",sep=""),width=5.25,height=3)
print(nzb_numt_corr_comp + theme(legend.position = "bottom") + guides(color =guide_legend(override.aes = list(size = 3)) ))
dev.off()

```

Exporting the contamination per condition:

```{r}
write.table(cond_nzb_est, file = paste(outdir_files,"nzb_condition_cont_est.txt", sep = ""), sep = "\t", quote = F, row.names = F)

```
